pkg/webhooks/hiveownership/hiveownership_test.go

```diff
@@ -0,0 +1,153 @@
+
+	"k8s.io/api/admission/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+type hiveOwnershipTestSuites struct {
+	testName        string
+	testID          string
+	username        string
+	userGroups      []string
+	oldObject       *runtime.RawExtension
+	operation       v1beta1.Operation
+	labels          map[string]string
+	shouldBeAllowed bool
+}
+
+const testObjectRaw string = `{
+	"metadata": {
+	  "name": "%s",
+	  "uid": "%s",
+	  "creationTimestamp": "2020-05-10T07:51:00Z",
+	  "labels": %s
+	},
+	"users": null
+  }`
+
+// labelsMapToString is a helper to turn a map into a JSON fragment to be
+// inserted into the testNamespaceRaw const. See createRawJSONString.
+func labelsMapToString(labels map[string]string) string {
+	ret, _ := json.Marshal(labels)
+	return string(ret)
+}
+
+func createRawJSONString(name, uid string, labels map[string]string) string {
+	return fmt.Sprintf(testObjectRaw, name, uid, labelsMapToString(labels))
+}
+func createOldObject(name, uid string, labels map[string]string) *runtime.RawExtension {
+	return &runtime.RawExtension{
+		Raw: []byte(createRawJSONString(name, uid, labels)),
+	}
+}
+
+func runTests(t *testing.T, tests []hiveOwnershipTestSuites) {
+	gvk := metav1.GroupVersionKind{
+		Group:   "quota.openshift.io",
+		Version: "v1",
+		Kind:    "ClusterResourceQuota",
+	}
+	gvr := metav1.GroupVersionResource{
+		Group:    "quota.openshift.io",
+		Version:  "v1",
+		Resource: "clusterresourcequotas",
+	}
+
+	for _, test := range tests {
+		obj := createOldObject(test.testName, test.testID, test.labels)
+		hook := NewWebhook()
+		httprequest, err := testutils.CreateHTTPRequest(hook.GetURI(),
+			test.testID,
+			gvk, gvr, test.operation, test.username, test.userGroups, obj, test.oldObject)
+		if err != nil {
+			t.Fatalf("Expected no error, got %s", err.Error())
+		}
+
+		response, err := testutils.SendHTTPRequest(httprequest, hook)
+		if err != nil {
+			t.Fatalf("Expected no error, got %s", err.Error())
+		}
+		if response.UID == "" {
+			t.Fatalf("No tracking UID associated with the response: %+v", response)
+		}
+
+		if response.Allowed != test.shouldBeAllowed {
+			t.Fatalf("Mismatch: %s (groups=%s) %s %s. Test's expectation is that the user %s",
+				test.username, test.userGroups,
+				testutils.CanCanNot(response.Allowed), string(test.operation),
+				testutils.CanCanNot(test.shouldBeAllowed))
+		}
+	}
+}
+
+func TestThing(t *testing.T) {
+	tests := []hiveOwnershipTestSuites{
+		{
+			testID:          "kube-admin-test",
+			username:        "kube:admin",
+			userGroups:      []string{"kube:system", "system:authenticated", "system:authenticated:oauth"},
+			operation:       v1beta1.Create,
+			shouldBeAllowed: true,
+		},
+		{
+			testID:          "dedi-admin-create-default",
+			username:        "ded-admin",
+			userGroups:      []string{adminGroups[0], "system:authenticated", "system:authenticated:oauth"},
```

Comment:
I think the names in this test case are a bit misleading. This isn't a dedicated-admin user because the `userGroups` is including a privileged group. If you do mean to test dedicated-admin, you don't want to include the adminGroups reference, because that's an SREP group. Instead, if you're testing dedicated-admins, you should include explicitly the `dedicated-admins` group. If you're intending to test an SREP member, the test ID and username should probably change.

Reply:
Yeah, that's true.. I've renamed the username and test-ID in that way.